### PR TITLE
[break] Remove space from prefixes

### LINKED ISSUE #129: Problems with prefix if no space after comment marker
Tried this with the sanitize file command
```java
//REPOBEE-SANITIZER-START
      return 1;
//REPOBEE-SANITIZER-REPLACE-WITH
      //return -1;
//REPOBEE-SANITIZER-END
```
This failed with 
```
sanitize-file: ERROR
Syntax errors detected in 1 file(s):

File.java
    Line 6: Missing prefix:             // 

```

After some debugging I found that the reason was that there was no space between `//` and `return -1;`. If this is intended, maybe make that clear in the docs?

### PR DESCRIPTION
**Pull request by tohanss**
Prefixes in replace blocks and on markers can now be played anywhere regardless of whitespace

Fix #129

### CODE REVIEW

**Comment by slarse** [comment]
Looks good at first glance, I want to try to break it a bit though. I'll try to get back to this during the weekend, been crunching too much at work the past few days to have after-work energy :)

**Review by slarse — CHANGES_REQUESTED** [review]
Implementation and functionality looks good to me! Nice work, this will make using sanitizer a LOT easier I think.

There are however two things I'd like to have done.

- [ ] Extend the documentation of prefixes to show this functionality, and why you'd want to use it (easy alignment of indentation). Clearly exemplify that only the prefix is removed, and not the preceding whitespace.
- [ ] There are a bunch of unrelated formatting changes both in this PR and in #137. They will conflict. Make a separate PR with those formatting changes, then merge that into these two to clean up the PRs.

**Comment by slarse** [comment]
Oh, and this is actually a breaking change, because before this whitespace counted as prefixes. So it should be marked `[break]`.

**Comment by tohanss** [comment]
That should just be removing the "format with black" commit, and creating a new branch with the formatting based of master right?

**Comment by slarse** [comment]
> That should just be removing the "format with black" commit, and creating a new branch with the formatting based of master right?

First of all just do a new PR from master with the formatting changes, and merge that into master. Then just merge master into your two feature PRs. You'll get a few merge conflicts but they'll be easy tor resolve.

File: README.md [hunk 1]
```diff
@@ -203,7 +203,11 @@ There are some rules for prefixing to observe:
 
 `Sanitizer`:
 
-* Determines prefix as any text (including whitespace) that comes before `REPOBEE-SANITIZER-START`
+* Determines prefix as any text before `REPOBEE-SANITIZER-START`
+  * This excludes any whitespace before the `START` marker
+    * Pro-Tip: There can be any amount of whitespace on any line using a prefix, see figure 4. This allows easy alignment of code
```
Comment (tohanss) [inline README.md:208]: I quite like having this as a "pro-tip" as it draws the eye of the reader towards what it really is, an explanation of a neat feature

**Comment by tohanss** [comment]
Should this have a [docs] tags as well as [break]?

**Comment by slarse** [comment]
> Should this have a [docs] tags as well as [break]?

No, just pick the one that's most important. Any feature change naturally has documentation changes as well, but the important part is the feature change.

File: README.md [hunk 2]
```diff
@@ -203,21 +202,25 @@ There are some rules for prefixing to observe:
...
 
-* Determines prefix as any text (including whitespace) that comes before `REPOBEE-SANITIZER-START`
-* Only determines prefix on a block-to-block basis, meaning that the prefix
+* Removes prefixes on `REPLACE-WITH` blocks when sanitizing
+  * Only the first occurrence of a prefix on each line is removed, allowing code comments to be preserved when sanitizing
+* Determines the prefix as any text before `REPOBEE-SANITIZER-START`
+  * This excludes any whitespace before the `START` marker
+    * Pro-Tip: There can be any amount of whitespace on any line using a prefix, see example 4. This allows easy alignment of code
+  * When the prefix is removed after running `Sanitizer`, all whitespace is preserved, only the prefix is removed
+* Determines prefix on a block-to-block basis, meaning that the prefix
```
Comment (slarse) [inline README.md:211]: This is very hard to read to me. I don't think using bullet point lists to present information this complex is appropriate. The bullet point list here prior to this PR was okayishly readable, but this is just way too much for my eyes. Have a look in the rendered document, can you see what I mean?

Although not really related to this PR (but this is the first time I see this text), I noticed that the text leading up to the bullet point list is also pretty strange, with two colon-terminated half-sentences like so:

```
There are some rules for prefixing to observe:

Sanitizer:
```

And not all of the bullet points can even be read as "Sanitizer <does X>", for example `Sanitizer: Code between replace and end markers ..."` doesn't make sense.

I think this whole thing would be much better described as fluent text describing how a prefix is defined, and then how it is used. Special cases could then potentially be documented in a bullet point list. For example, it could be structured something like so:

> Prefixes are valid for a single block, and are defined on the same line as the `REPOBEE-SANITIZER-START` marker of that block. All text starting from the first non-whitespace up until (but not including) the start marker counts as a prefix, and is valid for the rest of that block. Blabla blablabla blabla.

**Review by slarse — CHANGES_REQUESTED** [review]
Still happy with the implementation, but I'm not quite convinced by the docs. See the comment.

File: README.md (hunk 2 above)
Comment (tohanss) [reply README.md:211]: The indented points are related to its parent so it makes sense for me when I wrote it like that, but thinking about how others would perceive it I think you're right. Also i can agree that it's just too much right now, so ill rewrite it to text

File: repobee_sanitizer/_sanitize.py [hunk 3]
```diff
@@ -26,17 +26,22 @@ def sanitize_text(lines: List[str], strip: bool = False) -> Optional[str]:
...
+    prefix = ""
+    in_replace = False
     for line in lines:
         marker = _syntax.contained_marker(line)
         if marker == Markers.START:
-            prefix = re.match(rf"(.*?){Markers.START.value}", line).group(1)
-            prefix_length = len(prefix) if not strip else 0
+            prefix = (
+                re.match(rf"(.*?){Markers.START.value}", line).group(1).strip()
+            )
```
Comment (slarse) [inline repobee_sanitizer/_sanitize.py:36]: I didn't notice this before, but this should be a function `search_prefix` or something like that as it's duplicated in the syntax check.

File: README.md [hunk 4]
```diff
@@ -199,25 +198,20 @@ class StackTest {
...
 
-* Determines prefix as any text (including whitespace) that comes before `REPOBEE-SANITIZER-START`
-* Only determines prefix on a block-to-block basis, meaning that the prefix
-  selected at a `START` marker must be used until and including the next `END`
-  marker
-  * This means that all `Sanitizer` blocks can have individual
-    prefixes
-* Code between replace and end markers **MUST** also be prefixed if one is used
-* Prefixes inside `REPLACE-WITH` blocks are removed when sanitizing
+All lines inside a `REPLACE` block must have the prefix (if used), which is then stripped from the line. All whitespace is preserved when sanitizing and only the first occurrence of a prefix on each line is removed, allowing code comments inside the `REPLACE` block to be preserved.
```
Comment (slarse) [inline README.md:205]:
```suggestion
If a prefix is defined in a block, all lines in the remainder of the block that contain a marker, as well as all lines in `REPLACE` blocks, must contain the prefix before any other non-whitespace characters. There is no limit to the amount of whitespace that may appear before or after the prefix, however. All whitespace is also preserved when sanitizing, and only the first occurrence of a prefix on each line is removed, allowing code comments inside the `REPLACE` block to be preserved.
```

File: README.md [hunk 5]
```diff
@@ -199,25 +198,20 @@ class StackTest {
 markers. This way we can have java comments: `//`, python comments: `#` or
 similar preceding our markers. **This means code can still compile!**
 
-There are some rules for prefixing to observe:
+### There are some rules for prefixing to observe:
 
-`Sanitizer`:
+Prefixes are valid for a single block, and are defined on the same line as the `REPOBEE-SANITIZER-START` marker of that block. All text found before the `START` marker (stripped fom whitespace) counts as the prefix.
```
Comment (slarse) [inline README.md:203]:
```suggestion
Prefixes are valid for a single block, and are defined on the same line as the `REPOBEE-SANITIZER-START` marker of that block. The prefix is defined as all characters occurring before the `START` marker, without leading and trailing whitespace. Whitespace _within_ the prefix, such as in `/   /`, counts as part of the prefix.

File: README.md [hunk 6]
```diff
@@ -199,25 +198,20 @@ class StackTest {
...
-* Only determines prefix on a block-to-block basis, meaning that the prefix
-  selected at a `START` marker must be used until and including the next `END`
-  marker
-  * This means that all `Sanitizer` blocks can have individual
-    prefixes
-* Code between replace and end markers **MUST** also be prefixed if one is used
-* Prefixes inside `REPLACE-WITH` blocks are removed when sanitizing
+All lines inside a `REPLACE` block must have the prefix (if used), which is then stripped from the line. All whitespace is preserved when sanitizing and only the first occurrence of a prefix on each line is removed, allowing code comments inside the `REPLACE` block to be preserved.
+
+> Pro-Tip: There can be any amount of whitespace on any line using a prefix, see example 4. This allows easy alignment of code
```
Comment (slarse) [inline README.md:207]: I'm not sure I agree this is a pro tip, it kind of seems like _the_ way to use the replace block.

**Review by slarse — CHANGES_REQUESTED** [review]
Looking much better now. I've added some missing details.

As a personal preference, I'd suggest using parentheses a bit less, and especially don't put crucial information in them. A parenthesis suggests that what's written inside it is something of a side note, tangential to the topic at hand. So, for example, I would not put the fact that leading and trailing whitespace is ignored when determining a prefix, as that's crucial information.

There may also be cause to add a _summary_ in the form of a bullet point list. Maybe something like this.

```
To summarize, a prefix:

* Is defined by as all characters leading up to a `START` marker, without leading and trailing whitespace
* Is valid for the duration of the block it is defined in
* Must be the first non-whitespace sequence of characters on any line containing a marker
* Must be the first non-whitespace sequence of characters on all lines within a replace block
    - And when used in replace blocks, only the first instance of the prefix is removed
    - Whitespace (both leading and trailing) is preserved, making for easy code alignment!
```

File: README.md (hunk 4 above)
Comment (slarse) [reply README.md:205]: You missed that the marker lines also must have the prefix :)

File: README.md (hunk 5 above)
Comment (slarse) [reply README.md:203]: Note that I removed "stripped" here, because it's ambiguous. You even use it ambiguously within this PR. Here, it refers to removing leading and trailing whitespace, but when you talk about "stripping the prefix", it actually refers to just removing (has nothing to do with leading or trailing anything else).

File: README.md (hunk 6 above)
Comment (tohanss) [reply README.md:207]: Yeah, let's just remove it, people will figure that out, we also show it in the example
